### PR TITLE
Use std::map for main dictionary in ensemble_config

### PR DESCRIPTION
**Pull request by joakim-hove**
There are many `hash_type` references left in the ensemble_config.cpp file - but this is a start.

### CODE REVIEW

File: lib/enkf/ensemble_config.cpp [hunk 1]
```diff
@@ -191,8 +197,8 @@ void ensemble_config_free(ensemble_config_type * ensemble_config) {
 ert_impl_type ensemble_config_impl_type(const ensemble_config_type *ensemble_config, const char * ecl_kw_name) {
   ert_impl_type impl_type = INVALID;
 
-  if (hash_has_key(ensemble_config->config_nodes , ecl_kw_name)) {
-    enkf_config_node_type * node = (enkf_config_node_type *)hash_get(ensemble_config->config_nodes , ecl_kw_name);
+  if (ensemble_config->config_nodes.count(ecl_kw_name) > 0) {
+    const enkf_config_node_type * node = ensemble_config->config_nodes.at(ecl_kw_name);
```
Comment (andreabrambilla) [inline lib/enkf/ensemble_config.cpp]: in these situations where you have to check whether an object is present, and in that case do operations on it, it is more efficient to do it as follows
```
const auto node_it = ensemble_config->config_nodes.find(ecl_kw_name);
if(node_it != ensemble_config->config_nodes.end()) {
    const enkf_config_node_type * node = node_it.second;
    ...
```

File: lib/enkf/ensemble_config.cpp [hunk 2]
```diff
@@ -203,9 +209,9 @@ ert_impl_type ensemble_config_impl_type(const ensemble_config_type *ensemble_con
 
 enkf_var_type ensemble_config_var_type(const ensemble_config_type *ensemble_config, const char * ecl_kw_name) {
   enkf_var_type var_type = INVALID_VAR;
+  if (ensemble_config->config_nodes.count(ecl_kw_name) > 0) {
+    const enkf_config_node_type * node = ensemble_config->config_nodes.at(ecl_kw_name);
```
Comment (andreabrambilla) [inline lib/enkf/ensemble_config.cpp]: same

File: lib/enkf/ensemble_config.cpp [hunk 3]
```diff
@@ -216,14 +222,14 @@ enkf_var_type ensemble_config_var_type(const ensemble_config_type *ensemble_conf
...
+  return ensemble_config->config_nodes.count(key) > 0;
 }
 
 
 
 enkf_config_node_type * ensemble_config_get_node(const ensemble_config_type * ensemble_config, const char * key) {
-  if (hash_has_key(ensemble_config->config_nodes , key)) {
-    enkf_config_node_type * node = (enkf_config_node_type *)hash_get(ensemble_config->config_nodes , key);
+  if (ensemble_config->config_nodes.count(key) > 0) {
+    enkf_config_node_type * node = ensemble_config->config_nodes.at(key);
```
Comment (andreabrambilla) [inline lib/enkf/ensemble_config.cpp]: same

File: lib/enkf/ensemble_config.cpp [hunk 4]
```diff
@@ -232,11 +238,10 @@ enkf_config_node_type * ensemble_config_get_node(const ensemble_config_type * en
...
 
 enkf_config_node_type * ensemble_config_get_or_create_summary_node(ensemble_config_type * ensemble_config, const char * key) {
-    if (!hash_has_key(ensemble_config->config_nodes , key)) {
-        ensemble_config_add_summary(ensemble_config, key, LOAD_FAIL_SILENT);
-    }
+  if (ensemble_config->config_nodes.count(key) == 0)
+    ensemble_config_add_summary(ensemble_config, key, LOAD_FAIL_SILENT);
 
-    return ensemble_config_get_node(ensemble_config, key);
+  return ensemble_config_get_node(ensemble_config, key);
```
Comment (andreabrambilla) [inline lib/enkf/ensemble_config.cpp:246]: this is slightly different but same concept

File: lib/enkf/ensemble_config.cpp [hunk 5]
```diff
@@ -911,8 +908,8 @@ void ensemble_config_update_custom_kw_config(ensemble_config_type * config, cust
 enkf_config_node_type * ensemble_config_add_summary(ensemble_config_type * ensemble_config , const char * key , load_fail_type load_fail) {
   enkf_config_node_type * config_node = NULL;
 
-  if (hash_has_key(ensemble_config->config_nodes, key)) {
-    config_node = (enkf_config_node_type * ) hash_get(ensemble_config->config_nodes, key);
+  if (ensemble_config->config_nodes.count(key) > 0) {
+    config_node = ensemble_config->config_nodes.at(key);
```
Comment (andreabrambilla) [inline lib/enkf/ensemble_config.cpp]: and this is the last one

**Review by andreabrambilla — APPROVED** [review]
Overall the PR looks good :+1:

I have added a comment that affects several places. Some could argue that's good practice, other can say it's a minor details given that performance is probably not critical here. Change it if you'd like too, merge it as is if you have better things to do :wink:

**Comment by joakim-hove** [comment]
Thank you for the review - I have updated according to your suggestion.

**Comment by joakim-hove** [comment]
> merge it as is if you have better things to do

Thank - I would if I could; but someone else must push the green ...

**Comment by joakim-hove** [comment]
Observe that I can not merge this myself - cool if someone with the magical power will push the green. The same applies the TestArea pr.

File: lib/enkf/ensemble_config.cpp [hunk 6]
```diff
@@ -788,30 +791,26 @@ stringlist_type * ensemble_config_alloc_keylist_from_var_type(const ensemble_con
...
-  bool ret = false;
-  hash_iter_type * iter = hash_iter_alloc(config->config_nodes);
-  while (!hash_iter_is_complete( iter )) {
-    const char * key = hash_iter_get_next_key(iter);
-    if (enkf_config_node_get_impl_type( (enkf_config_node_type * ) hash_get(config->config_nodes , key)) == impl_type) {
-      ret = true;
-      break;
-    }
+  for (const auto& config_pair : config->config_nodes) {
+    const char * key       = config_pair.first.c_str();
```
Comment (markusdregi) [inline lib/enkf/ensemble_config.cpp]: Does not seem like you use the key for anything?

**Comment by markusdregi** [comment]
I had one comment, besides that I’m all good for merging this one!

File: lib/enkf/ensemble_config.cpp (hunk 6 above)
Comment (joakim-hove) [reply lib/enkf/ensemble_config.cpp]: Looks like you are right - fixed.

**Comment by markusdregi** [comment]
Good work!

**Pull request merged**